drivers/spi/spi_egis_et171.c

```diff
@@ -0,0 +1,973 @@
+						    current_rx->len);
+			configure_rx_dma_block_source(next_blk_cfg, data, cfg);
+
+			blk_cfg = next_blk_cfg;
+			next_blk_cfg->next_block = NULL;
+			remain_len -= ctx->current_rx->len;
+			spi_context_update_rx(ctx, dfs, ctx->current_rx->len);
+		}
+	} else {
+		data->dma_rx.dma_blk_cfg.next_block = NULL;
+		data->dma_rx.dma_cfg.block_count = 1;
+		data->dma_rx.dma_cfg.dma_callback = dma_rx_callback;
+	}
+
+	/* pass our client origin to the dma: data->dma_rx.channel */
+	ret = dma_config(data->dma_rx.dma_dev, data->dma_rx.channel,
+			&data->dma_rx.dma_cfg);
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		data->dma_rx.block_idx = 0;
+		data->dma_rx.dma_blk_cfg.next_block = NULL;
+		return ret;
+	}
+
+	return 0;
+}
+
+static int spi_transfer_dma(const struct device *dev)
+{
+	const struct spi_et171_cfg *const cfg = dev->config;
+	struct spi_et171_data * const data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	uint32_t data_len, tctrl, dma_rx_enable, dma_tx_enable;
+	int error = 0;
+
+	data_len = data->chunk_len - 1;
+
+	if (data_len > MAX_TRANSFER_CNT) {
+		return -EINVAL;
+	}
+
+	if (!spi_context_rx_on(ctx)) {
+		tctrl = (TRNS_MODE_WRITE_ONLY << TCTRL_TRNS_MODE_OFFSET);
+		dma_rx_enable = 0;
+		dma_tx_enable = 1;
+		sys_write32(data_len, SPI_WR_TRANS_CNT(cfg->base));
+	} else if (!spi_context_tx_on(ctx)) {
+		tctrl = (TRNS_MODE_READ_ONLY << TCTRL_TRNS_MODE_OFFSET);
+		dma_rx_enable = 1;
+		dma_tx_enable = 0;
+		sys_write32(data_len, SPI_RD_TRANS_CNT(cfg->base));
+	} else {
+		tctrl = (TRNS_MODE_WRITE_READ << TCTRL_TRNS_MODE_OFFSET);
+		dma_rx_enable = 1;
+		dma_tx_enable = 1;
+		sys_write32(data_len, SPI_WR_TRANS_CNT(cfg->base));
+		sys_write32(data_len, SPI_RD_TRANS_CNT(cfg->base));
+	}
+
+	sys_write32(tctrl, SPI_TCTRL(cfg->base));
+
+	/* Setting DMA config*/
+	error = spi_dma_move_buffers(dev);
+	if (error != 0) {
+		return error;
+	}
+
+	/* Enable END Interrupts */
+	sys_write32(IEN_END_MSK, SPI_INTEN(cfg->base));
+	/* Start transferring */
+	sys_write32(0, SPI_CMD(cfg->base));
+
+	if (dma_rx_enable) {
+		spi_rx_dma_enable(dev);
+		error = dma_start(data->dma_rx.dma_dev, data->dma_rx.channel);
+		if (error != 0) {
+			return error;
+		}
+	}
+	if (dma_tx_enable) {
+		spi_tx_dma_enable(dev);
+		error = dma_start(data->dma_tx.dma_dev, data->dma_tx.channel);
+		if (error != 0) {
+			return error;
+		}
+	}
+
+	return 0;
+}
+#endif
+
+static int transceive(const struct device *dev,
+			  const struct spi_config *config,
+			  const struct spi_buf_set *tx_bufs,
+			  const struct spi_buf_set *rx_bufs,
+			  bool asynchronous,
+			  spi_callback_t cb,
+			  void *userdata)
+{
+	const struct spi_et171_cfg *const cfg = dev->config;
+	struct spi_et171_data * const data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int error, dfs;
+	size_t chunk_len;
+
+	spi_context_lock(ctx, asynchronous, cb, userdata, config);
+	error = configure(dev, config);
+
+	if (error == 0) {
```

Comment:
Would be simpler to follow what's next by doing
if (error !=0)
        goto out;

it's more consistent with how you treat errors elsewhere (you test for an actual error, not a success)